fms/utils/generation.py

```diff
@@ -4,6 +4,91 @@
 import torch.nn.functional as F
 
 
+def __create_prefill_mask(
+    prompt_lengths: List[int], device: Union[str, torch.device]
+) -> torch.Tensor:
+    """Create a prefill mask based on prompt_lengths. When referring to prefill, this is typically the first iteration
+    before the kv-cache has had values stored in it for a given prompt
+
+    Args:
+        prompt_lengths: List[int]
+            list of prompt lengths
+        device: Union[str, torch.device]
+            device to put mask on
+
+    Returns:
+    torch.Tensor
+        a causal mask
+    """
+    max_tokens = max(prompt_lengths)
+
+    is_pad_list = []
+    for seq_len in prompt_lengths:
+        pads = torch.zeros(max_tokens - seq_len, dtype=torch.bool, device=device)
+        non_pads = torch.ones(seq_len, dtype=torch.bool, device=device)
+        is_pad_list.append(torch.cat((pads, non_pads)))
+    is_pad = torch.stack(is_pad_list)
+    mask = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+    return mask.tril()
+
+
+def __create_decode_mask(
+    context_lengths: List[int], device: Union[str, torch.device]
+) -> torch.Tensor:
+    """create a decode mask. When referring to decode, we are referring to all subsequent steps following prefill
+    (kv-cache has already been filled)
+
+    Args:
+        context_lengths: List[int]
+            current context length of each sequence in the batch
+        device: Union[str, torch.device]
+            device to put mask on
+
+    Returns:
+    torch.Tensor
+        a decode mask
```

Comment:
I think I mentioned this before but the different comment here seems to imply that a decode mask is something different from a causal mask, and I don't think it is?

the comments about these return types also don't add much value. if I needed detail of the return type, presumably I'd be looking for something like the dims or dtype or structure. if we're not providing any additional info, then we probably don't need them

Reply:
specifically, if I was checking the return type of a mask-getting function, I might be checking if it was a boolean mask or float mask (since sdpa supports both and handles them differently)